content/blog/2018/appswitch/appswitch.md

```diff
@@ -0,0 +1,81 @@
+
+Apart from introducing sufficiently long sleeps between service startups, a common pattern that is often used is to check for readiness of dependencies before starting a service.  In Kubernetes it is typically done as part of the init container of the pod.  However that means that the entire application would be held up until all its dependencies come alive.  Sometimes applications spend several minutes initializing themselves on startup before making their first outbound connection.  Not allowing the service to start adds substantial overhead to startup time.  Also, checking for readiness of dependencies won’t work in case of "multilevel" interdependencies where a service needs to be started before another service but only after a third service.  Likewise, multiple containers of the same pod may have mutual dependencies that cannot be ordered by simply waiting at the init container.
+
+### Example Scenario: IBM WebSphere ND
+
+Let us consider IBM WebSphere ND -- a widely deployed application middleware -- to grok these problems more closely.  It is a fairly complex framework in itself and consists of a central component called deployment manager (`dmgr`) that manages a set of node instances.  It uses UDP to negotiate cluster membership among the nodes and requires that deployment manager is up and operational before any of the node instances can come up and join the cluster.
+
+Why are we talking about a traditional application in the modern cloud-native context?  It turns out that there are significant gains to be had by enabling them to run on Kubernetes and Istio platforms.  Essentially it’s a part of the modernization journey that allows running traditional apps alongside green-field apps on the same modern platform to facilitate interoperation between the two.  As such, running WebSphere ND on Kubernetes requires a whole lot of additional cushion from AppSwitch in the way of the network environment it expects.  For the purpose of this blog however, I'll focus on the dependency ordering requirement.
+
+Simply deploying `dmgr` and node instances as pods on a Kubernetes cluster does not work.  `dmgr` and the node instances happen to have a lengthy initialization process that can take several minutes.  If they are all co-scheduled, the application typically ends up in a funny state.  When a node instance comes up and finds that `dmgr` is missing, it would take an alternate startup path.  Instead, if it had exited immediately, Kubernetes crash-loop would have taken over and perhaps the application would have come up.  But even in that case, it turns out a clean startup is not guaranteed.
```

Comment:
Why does the k8s crash loop logic not ensure a proper startup? Seems it should just work.